url70ed135da7_ralfbiedert_jcores_tgz-pJ8-sandbox_SimpleFilterTestJ8/src/net/jcores/jre/cores/CoreFile.java

```diff
@@ -0,0 +1,626 @@
+     * <br/>
+     * 
+     * @return The new core for the {@link AudioInputStream} objects.
+     */
+    public CoreAudioInputStream audio() {
+        return new CoreAudioInputStream(this.commonCore, map(new F1<File, AudioInputStream>() {
+            @Override
+            public AudioInputStream f(File x) {
+                return Sound.getStream(x);
+            }
+        }).array(AudioInputStream.class));
+    }
+
+    /**
+     * Copies all enclosed files to the destination. If <code>destination</code> is a directory all enclosed objects
+     * will be copied into that directory. If the destination is a file then it will be overwritten. In that case, and
+     * if this core encloses multiple files or directories, it is undefined what the content of <code>destination</code>
+     * will be afterwards<br/>
+     * <br/>
+     * 
+     * 
+     * Examples:
+     * <ul>
+     * <li><code>$("source.zip").file().copy("/dest/")</code> - Copies the file <code>source.zip</code> to the toplevel
+     * folder <code>/dest</code></li>
+     * <li><code>$("src/").file().copy("bin/")</code> - Copies the directory <code>src</code> to the directory
+     * <code>bin</code></li>
+     * </ul>
+     * 
+     * Multi-threaded.<br/>
+     * <br/>
+     * 
+     * @param destination The destination to write to. Can be a directory or a file. Directories <b>must end with a
+     * slash
+     * (<code>/</code>) or pre-exist</b>, otherwise they will be treated as files!
+     * 
+     * @return The new core file object, containing all files that have been copied..
+     */
+    public CoreFile copy(String destination) {
+        if (destination == null) {
+            this.commonCore.report(MessageType.MISUSE, "Destination null for copy().");
+            return this;
+        }
+
+        final File dest = new File(destination);
+        final CommonCore cc = this.commonCore;
+
+        return new CoreFile(this.commonCore, map(new F1<File, File[]>() {
+            @Override
+            public File[] f(File x) {
+                return Files.copy(cc, x, dest);
+            }
+        }).expand(File.class).array(File.class));
+    }
+
+    /**
+     * Opens the enclosed file streams as binary files and reads their data into byte
+     * buffers.
+     * File stream which could not be opened will be returned as null.<br/>
+     * <br/>
+     * 
+     * Examples:
+     * <ul>
+     * <li><code>$("data.jar").file().data().hash().print()</code> - Prints the (MD5) hash of the given file.</li>
+     * </ul>
+     * 
+     * Multi-threaded.<br/>
+     * <br/>
+     * 
+     * @return A CoreByteBuffer with binary content.
+     */
+    public CoreByteBuffer data() {
+        final CommonCore cc = this.commonCore;
+        return new CoreByteBuffer(this.commonCore, map(new F1<File, ByteBuffer>() {
+            public ByteBuffer f(File x) {
+                try {
+                    final FileChannel channel = new FileInputStream(x).getChannel();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *channel*. The resource is referenced at lines 260, 261, 264. The resource is closed at line 271. There are other execution paths that do not contain closure statements, for example, when an if condition at line 266 evaluates to true. Ensure *channel* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).